From 5365b8f100c6e9212bb25196f16ca0e1e8cc5c43 Mon Sep 17 00:00:00 2001
From: Dmitry Kazakov <dimula73@gmail.com>
Date: Tue, 6 Oct 2026 16:21:18 +0200
Subject: [PATCH] Fix QHeaderView failing to init memory mode

QHeaderViewPrivate::resizeSections() should bail out
if it was in low memory mode and it should stay in
low memory mode.

Hence, it must not begin to modify sections as if
it had section memory. That can lead to crashes
and other undefined behavior.

Related Krita bug:
https://bugs.kde.org/show_bug.cgi?id=519191

Pick-to: 6.11
Change-Id: I950ac68b9ceff99c583b40cb0221b602d6ac2165
Reviewed-by: Christian Ehrlicher <ch.ehrlicher@gmx.de>
Reviewed-by: Dmitry Kazakov <dimula73@gmail.com>
(cherry picked from commit e3d05b35b3a9ca95c19d8f8bba3e6f5e8b75ffa4)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
Origin: upstream, https://invent.kde.org/qt/qt/qtbase/-/commit/5365b8f100c6e9212bb25196f16ca0e1e8cc5c43
Forwarded: not-needed
Last-Update: 2026-10-29
---
 src/widgets/itemviews/qheaderview.cpp                |  5 ++---
 .../itemviews/qheaderview/tst_qheaderview.cpp        | 12 ++++++++++++
 2 files changed, 14 insertions(+), 3 deletions(-)

diff --git a/src/widgets/itemviews/qheaderview.cpp b/src/widgets/itemviews/qheaderview.cpp
index a666b79f6dc..5cd3d170b8e 100644
--- a/src/widgets/itemviews/qheaderview.cpp
+++ b/src/widgets/itemviews/qheaderview.cpp
@@ -3682,11 +3682,10 @@ void QHeaderViewPrivate::resizeSections(QHeaderView::ResizeMode globalMode, bool
     if (noSectionMemoryUsage() && (hasAutoResizeSections() || globalMode != QHeaderView::Fixed))
         setHeaderMode(HeaderMode::FlexibleWithSectionMemoryUsage);
 
-    if (sectionCount() == 0 )
+    // bail out on no section count, if recursive or if all sections should be same size (see above)
+    if (sectionCount() == 0 || resizeRecursionBlock || noSectionMemoryUsage())
         return;
 
-    if (resizeRecursionBlock)
-        return;
     resizeRecursionBlock = true;
 
     invalidateCachedSizeHint();
diff --git a/tests/auto/widgets/itemviews/qheaderview/tst_qheaderview.cpp b/tests/auto/widgets/itemviews/qheaderview/tst_qheaderview.cpp
index 834d5bced8a..ba302321179 100644
--- a/tests/auto/widgets/itemviews/qheaderview/tst_qheaderview.cpp
+++ b/tests/auto/widgets/itemviews/qheaderview/tst_qheaderview.cpp
@@ -215,6 +215,7 @@ private slots:
     void setSectionResizeModeWithSectionWillTakeMemory();
     void setModelWithAutoSizeWillSwitchToMemoryMode();
     void tableViewResizeSectionsWillSwitchToMemoryMode();
+    void noMemoryUsageOnNonResize();
 
     void setDefaultSectionSizeRespectsColumnWidth();
 
@@ -3808,5 +3809,16 @@ void tst_QHeaderView::setDefaultSectionSizeRespectsColumnWidth()
         QTRY_COMPARE(tree.columnWidth(c), columnWidths[c]);
 }
 
+void tst_QHeaderView::noMemoryUsageOnNonResize()
+{
+    TableViewWithBasicModel tv;
+    QVERIFY(tv.hasLowMemoryUsage());
+
+    // This should not change to memory usage.
+    tv.verticalHeader()->resizeSections(QHeaderView::Fixed);
+    QVERIFY(tv.hasLowMemoryUsage());
+}
+
+
 QTEST_MAIN(tst_QHeaderView)
 #include "tst_qheaderview.moc"
